README.md
# rrweb-stats


changelog.txt
11/3/2021:
stats.js:
	new methods:
	-shannonEntropy:
		inputs: list of probabilites, formula.
		return: shannon entropy value.
		description: calcualtes the shannon entropy of a list of probabilites.

rrwebDataMiner.js:
	new methods : 
		-findAbandonedForms:
			inputs : session.
			returns : from id and whether it was abandoned or not (object).
			description : gets all forms from a session and checks if they're abandoned or not.
		-_findForms:
			inputs : none.
			returns : form id and the id of all its children.
			description : gets all the forms and their children from the page.
		-_probabilityForEachClick:
			inputs : click interval.
			returns : probability for each element click (num of times element clicked/total nubmer of clicks).
			description: get the click probability for each element in a click interval.
		-isIntervalRageOrDeadClick:
			inputs : click interval, threshold.
			returns : enum intervalType.
			description : checks if the interval was rage clicks or dead clicks depends on the entropy threshold.
		-checkClickIntervals:
			inputs : entropy threshold, click threshold (default 0.5).
			returns : array similar to getClickIntervals, but instead of intervals, it will contain wether the interval was rage or dead clicks.
			description : moves through all the click intervals of every session and tells if they're rage or dead clicks.


4/3/2021:
-creating a new rrwebDataMiner no longer forces the user to enter the sessions at construction time.

-object now only storing sessions, wanted leaf types, leaves and the parsed data.

-various removals and rewrites.

-added new methods to rrwebDataMiner : 
	-getMovementIntervals:
		inputs : none.
		returns : array of movement intervals for each session.
		description : to get movement intervals.
		
	-getClickIntervals:
		inputs : timeThreshold(in seconds) - default: 0.5.
		returns : array of click intervals for each session.
		description : to get click intervals. 
		
	-getThrashingInfoOfMovInterval:
		inputs : movement, winSize, threshold.
		reutrns : object housing the probability, and thrashing period.
		description : to calcualte probability of thrashing for a single movement event.
		
	-getThrashingInfoOfSessionMovements:
		inputs : winSize, threshold.
		returns : an array of sessions, each containing its thrashing probabilites.
		description : goes through every movement in every session and runs "getThrashingInfoOfMovInterval" on it.
		
	-addSession:
		inputs : session, updateStats - default : true.
		return : void
		description : adds a session to the list of sessions, and updates stats depending on "updateStats".
		
	-addSessions:
		inputs : sessions, updateStats - default : true.
		return : void
		description : adds sessions to the list of sessions, and updates stats depending on "updateStats".
		
	-calculateStatstics:
		inputs : none.
		return : void
		description : to get the prased data, MUST be run before getting leaves, prased data or snapshot, or before calling any method using parsed/calculated data.
		

src/js/parse.js
import { rrwebDataMiner, eventTypes } from '/datapraser/rrwebDataMiner.js';
import { rebuild } from '/node_modules/rrweb-snapshot/es/rrweb-snapshot.js';

function displayData(data){
    let dataPoints = ["clicks", "hovers", "averageHoverTime", "visitorClickRate", "hoverToClickRate", "hoverToClickTime", "timeBeforeClick", "timeBeforeHover"]
    for(let dataPoint of dataPoints){
        document.getElementById(dataPoint).children[1].innerText = data[dataPoint];
    }
}

var dp;
fetch("/src/data/recordings.json")
    .then(response => response.json())
    .then(json => {
        dp = new rrwebDataMiner(json);
        dp.calculateStatstics();
        let iframe = document.getElementsByTagName("iframe")[0];
        let map = rebuild(dp.getSnapshot, {doc:iframe.contentDocument})[1];
        let head = iframe.contentDocument.getElementsByTagName("head")[0];
        let css = document.createElement("style")
        css.innerText = ".showing-data:hover {outline: 1px dashed blue;}";
        head.append(css);
        for(let leaf of dp.getLeaves){
            map[leaf].addEventListener("mouseenter", () => {
                displayData(dp.getElementData(leaf));
                map[leaf].classList.add("showing-data");
            });
            map[leaf].addEventListener("mouseleave", () => {
                map[leaf].classList.remove("showing-data");
            });
        }     
        console.log(dp.getClickIntervals(0.5));
        console.log(dp.checkClickIntervals(0.5, 0.5));
    })
    //code was run on a local server using local-server package in node
var formsdp = new rrwebDataMiner();
fetch("/src/data/form-recordings.json")
    .then(response => response.json())
    .then(json => {
        formsdp.addSessions(json, false);
        for(const session of json)
            console.log(formsdp.findAbandonedForms(session));
    })

datapraser/stats.js
class Stats{
    static sum(arr, formula){
        formula = (formula === undefined || formula === null) ? e => e: formula;
        let sum = 0;
        for(let index in arr) sum += formula(arr[index], arr[+index+1]);
        return sum;
    }

    static mean(arr, formula){
        formula = (formula === undefined || formula === null) ? e => e: formula;
        return this.sum(arr, formula)/arr.length;
    }

    static standardDeviation(arr, mean, formula){
        formula = (formula === undefined || formula === null) ? e => e: formula;
        mean = (mean === undefined || mean === null) ? this.mean(arr, formula) : mean;
        return Math.sqrt(this.sum(arr, e => Math.pow(formula(e) - mean, 2)) / arr.length);
    }

    static shannonEntropy(variables, formula){
        formula = (formula === undefined || formula === null) ? e => e: formula;
        //return this.sum(variables, e=> e * Math.log2(1/e));
        return -this.sum(variables, e => formula(e) * Math.log2(formula(e)));
    }
};

export {Stats};

datapraser/dataparser.js
import {IncrementalSource, MouseInteractions} from '/node_modules/rrweb/es/rrweb/src/types.js'
//replace it with relative place for the type.js from rrweb

function sortObject(object){
    return Object.keys(object).sort().reduce((r, k) => (r[k] = object[k], r), {});
}

function calculateTimeDifference(date1, date2){
    return (date2 - date1) /1000;
}

Object.defineProperty(Array.prototype, "sum", {
    value: function(){ return this.reduce((a, b)=>{return a+b}, 0); }
});

const eventTypes = {};

(function(eventTypes){
    eventTypes[eventTypes["documentLoad"] = 0] = "documentLoad";
    eventTypes[eventTypes["mouseEnter"] = 1] = "mouseEnter";
    eventTypes[eventTypes["mouseLeave"] = 2] = "mouseLeave";
    eventTypes[eventTypes["click"] = 3] = "click";
})(eventTypes);

class DataParser{
    constructor(config){
        
        this.defaultConfig = {
            sessions : null,
            leafTypes : [
                "input",
                "button",
                "a"
            ]
        }
        
        this.config = {...this.defaultConfig, ...config}
        
        if(!this.config.sessions || this.config.sessions.length === 0) throw new Error('Events must be provided at initlization');
        
        this.snapshot = this.config.sessions[0][3].data.node;
        this.leaves = this.findLeaves();
        this.organizedEvents = Array(this.config.sessions.length).fill().map(() => ({}));
        this.organizeEvents();
        this.parsedData = {}

        this.leaves.forEach(leaf => {
            this.parsedData[leaf] = {clicks:0, hovers:0, hoverAndClick:0, visitorClicks:0, clickTimes:[], hoverTimes:[], hoverToClickTimes:[], hoverDurTimes:[]};
        })
        
        this.parseEvents();
    }

    calculateOverLeaves(formula){
        let result = {};
        this.leaves.forEach(leaf => result[leaf] = formula(leaf));
        return result;
    }

    parseEvents(){
        for(let si = 0, session = this.organizedEvents[si]; si < Object.keys(this.organizedEvents).length; si++, session = this.organizedEvents[si]){
            let startTime;
            let alreadyClicked = [], alreadyHoverd = [];
            let lastEvent, lastMouseEnter;
            for(let key in session){
                let event = session[key];
                if(event.type == eventTypes.mouseEnter) {
                    this.parsedData[event.id].hovers++;
                    if(!alreadyHoverd.includes(event.id)){
                        this.parsedData[event.id].hoverTimes.push(calculateTimeDifference(startTime, key));
                        alreadyHoverd.push(event.id);
                    }
                }
                else if(event.type == eventTypes.click){
                    this.parsedData[event.id].clicks++;
                    if(lastEvent.event.type == eventTypes.mouseEnter) {
                        this.parsedData[event.id].hoverAndClick++;
                        this.parsedData[event.id].hoverToClickTimes.push(calculateTimeDifference(lastEvent.timestamp, key));
                        if(!alreadyClicked.includes(event.id)){
                            this.parsedData[event.id].clickTimes.push(calculateTimeDifference(startTime, key));
                            this.parsedData[event.id].visitorClicks++
                            alreadyClicked.push(event.id);
                        }
                    }
                }
                else if(event.type == eventTypes.mouseLeave){
                    this.parsedData[event.id].hoverDurTimes.push(calculateTimeDifference(lastMouseEnter.timestamp, key));
                }else if(event.type == 0) startTime = key;

                lastEvent = {event:session[key], timestamp:key};
                lastMouseEnter = event.type == eventTypes.mouseEnter ? {event:session[key], timestamp:key} : lastMouseEnter;
            }
        }
    }

    organizeEvents(){
        for(let si = 0, session = this.config.sessions[si]; si < this.config.sessions.length; si++, session = this.config.sessions[si]){
            let curData = this.organizedEvents[si];
            curData[session[3].timestamp] = {type:eventTypes.documentLoad};
            let lastHover = {id:-1, timestamp:-1};
            for(let i = 4, event = session[4]; i < session.length; i++, event = session[i]){
                if(event.data.source == IncrementalSource.MouseInteraction && event.data.type == MouseInteractions.Click && this.leaves.includes(event.data.id)) {curData[event.timestamp] = {id:event.data.id, type:eventTypes.click};}
                else if(event.data.source == IncrementalSource.MouseMove){
                    console.log(event);
                    event.data.positions.forEach(pos => {
                        if(lastHover.id != pos.id){
                            if(this.leaves.includes(lastHover.id)) curData[event.timestamp + pos.timeOffset] = {id:lastHover.id, type:eventTypes.mouseLeave};
                            if(this.leaves.includes(pos.id)) curData[ curData[event.timestamp + pos.timeOffset] ? event.timestamp + pos.timeOffset : event.timestamp + pos.timeOffset] = {id:pos.id, type:eventTypes.mouseEnter};
                            lastHover.id = pos.id;
                            lastHover.timestamp = event.timestamp + pos.timeOffset;
                        }
                    });
                } 
            }
            this.organizedEvents[si] = sortObject(this.organizedEvents[si]);
        }
    }

    findLeaves(){
        let leaves = [];
        let _this = this;

        function areChildrenOnlyTextNodes(node){
            for(let index in node.childNodes) if(node.childNodes[index].tagName) return false;
            return true;
        }
    
        function findLeavesRecursive(node){
            if(node.childNodes && node.childNodes.length > 0 && !areChildrenOnlyTextNodes(node)) node.childNodes.forEach( e => findLeavesRecursive(e));
            else if(node.tagName && _this.config.leafTypes.includes(node.tagName)) leaves.push(node.id);
        }
    
        findLeavesRecursive(this.snapshot);
        return leaves;
    }

    get getLeaves(){
        return this.leaves;
    }

    get hoverToClickRate(){
        return this.calculateOverLeaves(leaf => +(this.parsedData[leaf].hoverAndClick / this.parsedData[leaf].hovers * 100).toPrecision(2));
    }

    get visitorClickRate(){
        return this.calculateOverLeaves(leaf => +(this.parsedData[leaf].visitorClicks / this.config.sessions.length * 100).toPrecision(2));
    }

    get timeBeforeClick(){
        return this.calculateOverLeaves(leaf => +(this.parsedData[leaf].clickTimes.sum() / this.parsedData[leaf].clickTimes.length).toPrecision(2));
    }

    get timeBeforeHover(){
        return this.calculateOverLeaves(leaf => +(this.parsedData[leaf].hoverTimes.sum() / this.parsedData[leaf].hoverTimes.length).toPrecision(2));
    }

    get hoverToClickTime(){
        return this.calculateOverLeaves(leaf => +(this.parsedData[leaf].hoverToClickTimes.sum() / this.parsedData[leaf].hoverToClickTimes.length).toPrecision(2));
    }

    get averageHoverTime(){
        return this.calculateOverLeaves(leaf => +(this.parsedData[leaf].hoverDurTimes.sum() / this.parsedData[leaf].hoverDurTimes.length).toPrecision(2));
    }

    get getParsedData(){
        return this.parsedData;
    }

    get getCalculatedData(){
        let array = [
            this.visitorClickRate,
            this.averageHoverTime,
            this.hoverToClickRate,
            this.hoverToClickTime,
            this.timeBeforeClick,
            this.timeBeforeHover
        ];

        let returnObj = {};

        this.leaves.forEach(leaf => {
            returnObj[leaf] = {
                visitorClickRate: array[0][leaf],
                averageHoverTime: array[1][leaf],
                hoverToClickRate: array[2][leaf],
                hoverToClickTime: array[3][leaf],
                timeBeforeClick: array[4][leaf],
                timeBeforeHover: array[5][leaf]
            };
        })

        
        return returnObj;
    }

    get getSnapshot(){
        return this.snapshot;
    }
};

export {DataParser, eventTypes};

datapraser/rrwebDataMiner.js
import {IncrementalSource, MouseInteractions, EventType} from '/node_modules/rrweb/es/rrweb/src/types.js'
import {Stats} from './stats.js'

function sortObject(object){
    return Object.keys(object).sort().reduce((r, k) => (r[k] = object[k], r), {});
}

function calculateTimeDifference(date1, date2){
    return (date2 - date1) /1000;
}

const eventTypes = {};
(function(eventTypes){
    eventTypes[eventTypes["documentLoad"] = 0] = "documentLoad";
    eventTypes[eventTypes["mouseEnter"] = 1] = "mouseEnter";
    eventTypes[eventTypes["mouseLeave"] = 2] = "mouseLeave";
    eventTypes[eventTypes["click"] = 3] = "click";
})(eventTypes);

const intervalType = {};
(function(intervalType){
    intervalType[intervalType["rageClick"] = 0] = "rageClick";
    intervalType[intervalType["deadClick"] = 1] = "deadClick";
})(intervalType);

class rrwebDataMiner{
    constructor(sessions, leafTypes){
        
        this.sessions = [];

        if(sessions) this.addSessions(sessions, false);

        this.leafTypes = leafTypes || ["input", "button", "a"];
        
        this.leaves = null;
        this.parsedData = {}

    }

    _clearParsedData(){
        this.parsedData = {}

        this.leaves.forEach(leaf => {
            this.parsedData[leaf] = {clicks:0, hovers:0, hoverAndClick:0, visitorClicks:0, clickTimes:[], hoverTimes:[], hoverToClickTimes:[], hoverDurTimes:[]};
        })
    }

    _calculateOverLeaves(formula){
        let result = {};
        this.leaves.forEach(leaf => result[leaf] = formula(leaf));
        return result;
    }

    _parseSessionEvents(organizedSession){
        let startTime;
        let alreadyClicked = [], alreadyHoverd = [];
        let lastEvent, lastMouseEnter;
        for(let key in organizedSession){
            let event = organizedSession[key];
            if(event.type == eventTypes.mouseEnter) {
                this.parsedData[event.id].hovers++;
                if(!alreadyHoverd.includes(event.id)){
                    this.parsedData[event.id].hoverTimes.push(calculateTimeDifference(startTime, key));
                    alreadyHoverd.push(event.id);
                }
            }
            else if(event.type == eventTypes.click){
                this.parsedData[event.id].clicks++;
                if(lastEvent.event.type == eventTypes.mouseEnter) {
                    this.parsedData[event.id].hoverAndClick++;
                    this.parsedData[event.id].hoverToClickTimes.push(calculateTimeDifference(lastEvent.timestamp, key));
                    if(!alreadyClicked.includes(event.id)){
                        this.parsedData[event.id].clickTimes.push(calculateTimeDifference(startTime, key));
                        this.parsedData[event.id].visitorClicks++
                        alreadyClicked.push(event.id);
                    }
                }
            }
            else if(event.type == eventTypes.mouseLeave){
                this.parsedData[event.id].hoverDurTimes.push(calculateTimeDifference(lastMouseEnter.timestamp, key));
            }else if(event.type == 0) startTime = key;
            lastEvent = {event:organizedSession[key], timestamp:key};
            lastMouseEnter = event.type == eventTypes.mouseEnter ? {event:organizedSession[key], timestamp:key} : lastMouseEnter;
        }
    }

    _parseSessionsEvents(organizedEvents){
        for(let si = 0, session = organizedEvents[si]; si < organizedEvents.length; si++, session = organizedEvents[si]){
            this._parseSessionEvents(session);
        }
    }

    _organizeSessionEvents(session){
        let organizedEvents = {};
        organizedEvents[session[3].timestamp] = {type:eventTypes.documentLoad};
        let lastHover = {id:-1, timestamp:-1}

        for(const event of session){
            if(event.data.source == IncrementalSource.MouseInteraction && event.data.type == MouseInteractions.Click && this.leaves.includes(event.data.id)) organizedEvents[event.timestamp] = {id:event.data.id, type:eventTypes.click};
            else if(event.data.source == IncrementalSource.MouseMove){
                event.data.positions.forEach(pos => {
                    if(lastHover.id != pos.id){
                        if(this.leaves.includes(lastHover.id)) organizedEvents[event.timestamp + pos.timeOffset] = {id:lastHover.id, type:eventTypes.mouseLeave};
                        if(this.leaves.includes(pos.id)) organizedEvents[ organizedEvents[event.timestamp + pos.timeOffset] ? event.timestamp + pos.timeOffset : event.timestamp + pos.timeOffset] = {id:pos.id, type:eventTypes.mouseEnter};
                        lastHover.id = pos.id;
                        lastHover.timestamp = event.timestamp + pos.timeOffset;
                    }
                });
            }
        }

        organizedEvents = sortObject(organizedEvents);

        return organizedEvents;
    }

    _organizeSessionsEvents(){
        let organizedSessionsEvents = [];
        for(const session of this.sessions) organizedSessionsEvents.push(this._organizeSessionEvents(session));
        return organizedSessionsEvents;
    }

    _findLeaves(){
        let leaves = [];
        let _this = this;

        function areChildrenOnlyTextNodes(node){
            for(const child of node.childNodes) if(child.tagName) return false;
            return true;
        }
    
        function findLeavesRecursive(node){
            if(node.childNodes && node.childNodes.length > 0 && !areChildrenOnlyTextNodes(node)) node.childNodes.forEach( e => findLeavesRecursive(e));
            else if(node.tagName && _this.leafTypes.includes(node.tagName)) leaves.push(node.id);
        }
    
        findLeavesRecursive(this.getSnapshot);
        return leaves;
    }

    _findForms(){
        let forms = {};

        function getFormChildren(form, node){
            form.add(node.id);
            if(node.childNodes) node.childNodes.forEach(child => getFormChildren(form, child));
        }

        function findFormsRecursive(node){
            if(node.tagName)
                if(node.tagName === "form") {
                    forms[node.id] = new Set()
                    getFormChildren(forms[node.id], node)
                }
                else node.childNodes.forEach(child => findFormsRecursive(child));
        }

        findFormsRecursive(this.getSnapshot.childNodes[1]);
        return forms;
    }

    _probabilityForEachClick(interval){
        let elementCount = {};
        let probablities = {};
        for(const clickEvent of interval){
            if(elementCount.hasOwnProperty(clickEvent.data.id)) elementCount[clickEvent.data.id]++;
            else{
                elementCount[clickEvent.data.id] = 0;
            }
        }
        
        for(const element in elementCount) probablities[element] = elementCount[element] / interval.length;
        return probablities;
    }

    isIntervalRageOrDeadClick(interval, threshold){
        return Stats.shannonEntropy(this._probabilityForEachClick(interval)) <= threshold ? intervalType.deadClick : intervalType.rageClick;
    }

    checkClickIntervals(entropyThreshold, clickThreshold = 0.5){
        return this.getClickIntervals(clickThreshold).map(e => e.map(a => this.isIntervalRageOrDeadClick(a, entropyThreshold)))
    }

    findAbandonedForms(session){
        if(!session) throw new Error("Please provide a correct session");

        let forms = this._findForms();
        let abandonedForms = {};

        for(const key in forms) abandonedForms[key] = true;

        function anyAbandoned(){
            for(const form in abandonedForms) 
                if(abandonedForms[form] === true) return true;
            return false; 
        }

        for(const event of session){
            if(event.type === EventType.IncrementalSnapshot && event.data.source !== IncrementalSource.MouseMove){
                for(const form in abandonedForms)
                    if(forms[form].has(event.data.id)) abandonedForms[form] = false;
                if(!anyAbandoned()) return abandonedForms;
            }
        }

        return abandonedForms;
    }

    getMovementIntervals(){
        function getMovementSpeed(movementEvent) {
            function calculateDistance(x1, y1, x2 ,y2){
                return Math.sqrt(Math.pow(x2 - x1, 2) + Math.pow(y2 - y1, 2));
            }

            function calculateMovementDistance(movement1, movement2) {
                return calculateDistance(movement1.x, movement1.y, movement2.x, movement2.y);
            }

            function getMoveTime(event, movement) {
                return movement.timeOffset == undefined ? movement.timestamp : event.timestamp + movement.timeOffset;
            }

            let positions = movementEvent.data.positions;
            let distanceSum = Stats.sum(positions, (n1, n2) => {
                if (n2 != undefined) return calculateMovementDistance(n1, n2);
                else return 0;
            });

            let date1 = getMoveTime(movementEvent, positions[0]);
            let date2 = getMoveTime(movementEvent, positions[positions.length - 1]);
            return positions.length > 1 ? {
                speed: Math.round(distanceSum / calculateTimeDifference(date1, date2)),
                start: date1,
                end: date2
            } : {speed:0, start:0, end:0};
        }

        let movementSpeeds = [];

        for(const session of this.sessions){
            let lastMovement = null;
            let tempMove = [];
            let sessionMoves = [];
            for(const event of session){
                if(event.data.source == IncrementalSource.MouseMove){
                    if (lastMovement != null){
                        let result = getMovementSpeed(event);
                        if(result.speed) tempMove.push(result);
                    } 
                    lastMovement = event;
                }
                if(event.data.source == IncrementalSource.MouseInteraction){
                    if(tempMove.length > 0){ 
                        sessionMoves.push(tempMove); 
                    }
                    tempMove = [];
                }
            }
            movementSpeeds.push(sessionMoves.length ? sessionMoves : [tempMove]);
        }

        return movementSpeeds;
    }

    getClickIntervals(timeThreshold = 0.5){
        let intervals = [];
        
        for(const session of this.sessions){
            let lastClick = null;
            let interval = [];
            intervals.push([]);
            
            for(const event of session){
                if(event.data.source == IncrementalSource.MouseInteraction && event.data.type == MouseInteractions.Click){
                    if(lastClick && calculateTimeDifference(lastClick.timestamp, event.timestamp) <= timeThreshold) {
                        if(interval.length === 0) interval.push(lastClick)
                        interval.push(event);
                        lastClick = event;
                    }
                    else if(interval.length > 0){
                        intervals[intervals.length - 1].push(interval);
                        interval = [];
                    }
                    lastClick = event;
                }
            }

            if(interval.length > 0) intervals[intervals.length - 1].push(interval);
        }

        return intervals;
    }

    getThrashingInfoOfMovInterval(movements, winSize, threshold) {
        if(movements.length >= winSize){
            var maxTrashingProbability = -1;
            var thrashingPeriod = [];

            for(let i = 0; i < movements.length; ++i) {
                var sum = 0;
                for (let j = i; j < movements.length && j-i < winSize; j++) {
                    var relSpeed = movements[j].speed / threshold;
                    var sigmoid = 1 / (1 + Math.exp(-Math.pow(relSpeed, 3)));
                    sum += sigmoid;
                }
                var avgSigmoid = (sum / winSize);
                if(avgSigmoid > 0.6) thrashingPeriod.push(movements[i]);
                if(maxTrashingProbability < avgSigmoid) maxTrashingProbability = avgSigmoid;
            }
            return {"thrashingProbability": maxTrashingProbability, "thrashingPeriod": thrashingPeriod} ;
        }
        return {"thrashingProbability": 0, "thrashingPeriod": null};
    }

    getThrashingInfoOfSessionMovements(winSize, threshold){
        let sessions = this.getMovementIntervals();
        let sessionsThrashings = [];
        for(const session of sessions){
            let sessionThrashings = [];
            for(const movement of session){
                sessionThrashings.push(this.getThrashingInfoOfMovInterval(movement, winSize, threshold));
            }
            sessionsThrashings.push(sessionThrashings);
        }
        return sessionsThrashings;
    }

    addSession(session, updateStats = true){
        this.sessions.push(session);
        if(updateStats) this._parseSessionEvents(this._organizeSessionEvents(session));
    }

    addSessions(sessions, updateStats = true){
        for(const session of sessions){
            this.addSession(session, updateStats);
        }
    }

    calculateStatstics(){
        this.leaves = this._findLeaves();
        this._clearParsedData()
        this._parseSessionsEvents(this._organizeSessionsEvents(this.sessions));
    }

    get getLeaves(){
        return this.leaves;
    }

    get getParsedData(){
        return this.parsedData;
    }

    get getCalculatedData(){
        let returnObj = {};

        this.leaves.forEach(leaf => {
            returnObj[leaf] = this.getElementData(leaf);
        });
        
        return returnObj;
    }

    getElementData(id){
        if(!this.leaves.includes(id)) throw new Error("Unknown leaf ID");
        return {
            ...this.parsedData[id],
            averageHoverTime: +(Stats.mean(this.parsedData[id].hoverDurTimes)).toPrecision(2),
            visitorClickRate: +(this.parsedData[id].visitorClicks / this.sessions.length * 100).toPrecision(2),
            hoverToClickRate: +(this.parsedData[id].hoverAndClick / this.parsedData[id].hovers * 100).toPrecision(2),
            hoverToClickTime: +(Stats.mean(this.parsedData[id].hoverToClickTimes)).toPrecision(2),
            timeBeforeClick: +(Stats.mean(this.parsedData[id].clickTimes)).toPrecision(2),
            timeBeforeHover: +(Stats.mean(this.parsedData[id].hoverTimes)).toPrecision(2)
        };
    }

    get getSnapshot(){
        return this.sessions[0][3].data.node;
    }

    get speeds(){
        return this.movementSpeeds;
    }
};

export {rrwebDataMiner, eventTypes, intervalType};